pandas/core/generic.py

```diff
@@ -5228,6 +5229,19 @@ def __setattr__(self, name: str, value) -> None:
         After regular attribute access, try setting the name
         This allows simpler access to columns for interactive use.
         """
+        # Fix for handling assignment to a Catagorical dtype
+        # with a category which is not predefined.
+        if isinstance(value, BlockManager) and is_categorical_dtype(self):
+            if len(value) > 0:
+                new_value = value.as_array()[-1]
+            else:
+                new_value = None
+            if not pd.isna(new_value) and new_value not in self.dtype.categories.values:
+                raise ValueError(
+                    "Cannot setitem on a Categorical with a new "
+                    "category, set the categories first"
+                )
+
```

Comment:
not sure this is the correct location for the fix.

I suspect _setitem_with_indexer in pandas\core\indexing.py should be changed as this function is for setting values on a Series/DataFrame

not keen on seeing is_categorical_dtype in the code. would prefer using _can_hold_element but that maybe a more reaching change.

Reply:
Thanks @simonjayhawkins! I'm not very familiar with the entire codebase, so your feedback has been very helpful to me. I'm currently having another look at it given your suggestions.